st2common/st2common/services/executions.py

```diff
@@ -135,18 +140,15 @@ def create_execution_object(liveaction, publish=True):
     attrs['log'] = [_create_execution_log_entry(liveaction['status'])]
 
     execution = ActionExecutionDB(**attrs)
-    execution = ActionExecution.add_or_update(execution, publish=False)
-
-    # Update the web_url field in execution. Unfortunately, we need
-    # the execution id for constructing the URL which we only get
-    # after the model is written to disk.
+    # TODO: Do 100% research this is fully safe and unique in distributed setups
+    execution.id = ObjectId()
     execution.web_url = _get_web_url_for_execution(str(execution.id))
     execution = ActionExecution.add_or_update(execution, publish=publish)
 
-    if parent:
-        if str(execution.id) not in parent.children:
-            parent.children.append(str(execution.id))
-            ActionExecution.add_or_update(parent)
+    if parent and str(execution.id) not in parent.children:
+        values = {}
+        values['push__children'] = str(execution.id)
```

Comment:
This should be more efficient and unlike previous approach (full document update), it's also atomic.